pkg/gateway/operations/putobject.go

```diff
@@ -32,75 +32,132 @@ const (
+func (controller *PutObject) RequiredPermissions(req *http.Request, repoID, _, destPath string) (permissions.Node, error) {
 	copySource := req.Header.Get(CopySourceHeader)
-	copySourceDecoded, err := url.QueryUnescape(copySource)
-	if err != nil {
-		copySourceDecoded = copySource
-	}
 
-	if len(copySourceDecoded) == 0 {
-		return &auth.OnePermission{
-			Action:   permissions.WriteObjectAction,
-			Resource: permissions.ObjectArn(repoID, destPath),
+	if len(copySource) == 0 {
+		return permissions.Node{
+			Permission: permissions.Permission{
+				Action:   permissions.WriteObjectAction,
+				Resource: permissions.ObjectArn(repoID, destPath)},
 		}, nil
 	}
-	// check this is a copy operation
-	p, err := path.ResolveAbsolutePath(copySourceDecoded)
+	// this is a copy operation
+	p, err := path.ResolveAbsolutePath(copySource)
 	if err != nil {
 		logging.Default().WithError(err).Error("could not parse copy source path")
-		return nil, gatewayErrors.ErrInvalidCopySource
+		return permissions.Node{}, gatewayErrors.ErrInvalidCopySource
 	}
-	return &auth.AndPermission{
-		&auth.OnePermission{
-			Action:   permissions.WriteObjectAction,
-			Resource: permissions.ObjectArn(repoID, destPath),
-		},
-		&auth.OnePermission{
-			Action:   permissions.ReadObjectAction,
-			Resource: permissions.ObjectArn(p.Repo, p.Path),
-		},
-	}, nil
+
+	return permissions.Node{
+		Type: permissions.NodeTypeAnd,
+		Nodes: []permissions.Node{
+			{
+				Permission: permissions.Permission{
+					Action:   permissions.WriteObjectAction,
+					Resource: permissions.ObjectArn(repoID, destPath)},
+			},
+			{
+				Permission: permissions.Permission{
+					Action:   permissions.ReadObjectAction,
+					Resource: permissions.ObjectArn(p.Repo, p.Path)},
+			}}}, nil
 }
 
+// extractEntryFromCopyReq: update metadata to refer to the source hash in the destination workspace
 func extractEntryFromCopyReq(w http.ResponseWriter, req *http.Request, o *PathOperation, copySource string) *catalog.DBEntry {
-	copySourceDecoded, err := url.QueryUnescape(copySource)
+	p, err := getPathFromSource(copySource)
 	if err != nil {
-		copySourceDecoded = copySource
+		o.Log(req).WithError(err).Error("could not parse copy source path")
+		_ = o.EncodeError(w, req, gatewayErrors.Codes.ToAPIErr(gatewayErrors.ErrInvalidCopySource))
+		return nil
 	}
-	p, err := path.ResolveAbsolutePath(copySourceDecoded)
+	ent, err := o.Catalog.GetEntry(req.Context(), o.Repository.Name, p.Reference, p.Path, catalog.GetEntryParams{})
+	if err != nil {
+		o.Log(req).WithError(err).Error("could not read copy source")
+		_ = o.EncodeError(w, req, gatewayErrors.Codes.ToAPIErr(gatewayErrors.ErrInvalidCopySource))
+		return nil
+	}
+	return ent
+}
+
+// extractAndCopyEntryFromCopyReq: create copy of the file
+func extractAndCopyEntryFromCopyReq(w http.ResponseWriter, req *http.Request, o *PathOperation, copySource string) *catalog.DBEntry {
```

Comment:
Doesn't really extract, this is basically a copy command that returning, please change name and documentation to reflect that